text/0000-less-unwinding.md

```diff
@@ -0,0 +1,281 @@
+- Feature Name: `panic_runtime`
+- Start Date: 2016-02-25
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Stabilize implementing panics as aborts.
+
+* Stabilize the `-Z no-landing-pads` flag under the name `-C panic=strategy`
+* Implement a number of unstable features akin to custom allocators to swap out
+  implementations of panic just before a final product is generated.
+* Add a `[profile.dev]` option to Cargo to configure how panics are implemented.
+
+# Motivation
+[motivation]: #motivation
+
+Panics in Rust have long since been implemented with the intention of being
+caught at particular boundaries (for example the thread boundary). This is quite
+useful for isolating failures in Rust code, for example:
+
+* Servers can avoid taking down the entire process but can instead just take
+  down one request.
+* Embedded Rust libraries can avoid taking down the entire process and can
+  instead gracefully inform the caller that an internal logic error occurred.
+* Rust applications can isolate failure from various components. The classical
+  example of this is Servo can display a "red X" for an image which fails to
+  decode instead of aborting the entire browser or killing an entire page.
+
+While these are examples where a recoverable panic is useful, there are many
+applications where recovering panics is undesirable or doesn't lead to anything
+productive:
+
+* Rust applications which use `Result` for error handling typically use `panic!`
+  to indicate a fatal error, in which case the process *should* be taken down.
+* Many applications simply can't recover from an internal assertion failure, so
+  there's no need trying to recover it.
+* To implement a recoverable panic, the compiler and standard library use a
+  method called stack unwinding. The compiler must generate code to support this
+  unwinding, however, and this takes time in codegen and optimizers.
+* Low-level applications typically don't use unwinding at all as there's no
+  stack unwinder (e.g. kernels).
+
+> **Note**: as an idea of the compile-time and object-size savings from
+> disabling the extra codegen, compiling Cargo as a library is 11% faster (16s
+> from 18s) and 13% smaller (15MB to 13MB). Sizable gains!
+
+Overall, the ability to recover panics is something that needs to be decided at
+the application level rather than at the language level. Currently the compiler
```

Comment:
At the application level can make sense, but what about libraries that want to opt-out because e.g. they can't recover from assertions. But what if you combine such library with applications that do want recoverable panics because they don't want to take the whole process down?